### PR TITLE
Produce errors when parsing a media type or manifest fails

### PR DESCRIPTION
**Pull request by kzys**
*Issue #, if available:*

#18 

*Description of changes:*

Changed the signature of parseImageManifestMediaType to return an error separately

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jahkeup — COMMENTED** [review]
Thanks @kzys, happy to see the mediaType parsing getting some improvements :muscle: 

Got a couple questions and small suggestions inlined.

File: ecr/resolver.go
Comment (kzys) [reply ecr/resolver.go:245]: I was following what was there before, but I do agree that is misleading for callers.

I think `return "", err` is appropriate. Introducing a special value (unparsedMediaType) means that clients need to check the value even err is not-nil, which is not ideal.

File: ecr/resolver.go
Comment (kzys) [reply ecr/resolver.go:273]: I see. How about just using `fmt.Errorf()` here? I'd like to include the version number into the error string.

File: ecr/resolver_test.go
Comment (kzys) [reply ecr/resolver_test.go:54]: Sure. I will update the test.

File: ecr/resolver_test.go
Comment (kzys) [reply ecr/resolver_test.go:55]: I'm unsure errors.Is() is safe to call against nil. Let me check.

Comment (kzys) [reply ecr/resolver_test.go]: It is safe to call against nil. Updated in the latest revision.

File: ecr/resolver.go
Comment (jahkeup) [reply ecr/resolver.go]: Yeah, including the unhandled version seems like a good idea to me. Can you wrap it with a [go113 style error](https://blog.golang.org/go1.13-errors) for use with `errors.Is` downstream?

File: ecr/resolver.go
Comment (jahkeup) [reply ecr/resolver.go]: Yeah, this is roughly what we noticed during a prior PR as well. I'm okay with propagating the error. Can we use a [go113 style error](https://blog.golang.org/go1.13-errors) here for distinction, eg: `ErrInvalidManifest` (or something like that)? We might want to use that to retain the existing fallback behavior in the updated call site.. I'll let @samuelkarp chime in regarding keeping it similar to what's there now.

**Comment by jahkeup** [comment]
Updated the title to reflect the current solution's approach to returned errors.

File: ecr/resolver.go
Comment (kzys) [reply ecr/resolver.go]: Thanks. Added `ErrInvalidManifest` in the last revision.

File: ecr/resolver.go
Comment (kzys) [reply ecr/resolver.go]: Sure. Added `ErrInvalidManifest` here as well.

File: ecr/resolver.go
Comment (jahkeup) [inline ecr/resolver.go:40]: Looks like an extra key press snuck in here.

```suggestion
	ErrInvalidManifest = errors.New("invalid manifest")
```

File: ecr/resolver.go
Comment (jahkeup) [inline ecr/resolver.go:285]: This should say something along the lines of `"unsupported schema version %d"` rather than appending the full manifest blob. The debug log level will already have included the manifest body earlier on and I'd prefer to see terse error strings used where possible.

**Review by jahkeup — CHANGES_REQUESTED** [review]
Couple small things and then I think this'll be ready for final review + merge (I want to chat with Sam before merge though).

File: ecr/resolver.go
Comment (kzys) [reply ecr/resolver.go]: Oops. Fixed in the last revision.

File: ecr/resolver.go
Comment (kzys) [reply ecr/resolver.go]: I see. Fixed in the last revision.

File: ecr/resolver.go
Comment (jahkeup) [inline ecr/resolver.go:193]: I can't think of regression cases that might be affected. I'm somewhat concerned that there's one I'm not thinking of right now that @samuelkarp had in mind when we last talked and when the behavior was first added. 

At first blush, it *seems* like we could be preventing alternative mediaType blobs from being transparently passed through (unexpected types, such as non-container image manifests/manifest-lists; eg: [artifacts](https://github.com/opencontainers/artifacts/tree/ecce2ed872459652b173fd5b5f253cdc55abf736)).. but then we only handle that case here if ECR doesn't present the manifest's mediaType in the service's response.

I'll reach out and check in through multiple channels.

File: ecr/resolver_test.go
Comment (jahkeup) [inline ecr/resolver_test.go:59]: nit: this can be an assert as well - the test cases expects that there's not an error *and* that the mediaType is as expected. There's no "NPE" we might hit on the way, so may as well present all the unmet expectations for a given failure rather than terminating earlier on.

**Review by jahkeup — APPROVED** [review]
Looks great! Thanks again @kzys :+1: 

I'll check in with @samuelkarp to verify that we're not regressing before merge.

**Review by samuelkarp — APPROVED** [review]
Thanks @kzys! My apologies for the (very) delayed review!

**Pull request merged**
